packages/basic-crawler/src/internals/basic-crawler.ts

```diff
@@ -1135,10 +1141,22 @@ export class BasicCrawler<Context extends CrawlingContext = BasicCrawlingContext
         if (shouldRetryRequest) {
             this.stats.errorTrackerRetry.add(error);
 
-            await this._tagUserHandlerError(() => this.errorHandler?.(this._augmentContextWithDeprecatedError(crawlingContext, error), error));
+            if (error instanceof SessionError) {
+                if ((request.sessionRotationCount ?? 0) > 10) {
+                    throw new Error(`Session rotation failed ${request.sessionRotationCount} times. `
+                        + 'This might be caused by a misconfigured proxy or an invalid session pool configuration.');
+                }
+                request.sessionRotationCount ??= 0;
+                request.sessionRotationCount++;
+                crawlingContext.session?.retire();
+            } else {
+                await this._tagUserHandlerError(() => this.errorHandler?.(this._augmentContextWithDeprecatedError(crawlingContext, error), error));
+            }
 
             if (!request.noRetry) {
-                request.retryCount++;
+                if (!(error instanceof SessionError)) {
+                    request.retryCount++;
+                }
```

Comment:
Maybe too much of a patch - do we have a way to retry a request without increasing the retry count?

Reply:
We have the `RetryRequestError`, if you throw that, it will be always retried, not sure if that's enough for this use case.

Reply:
Yep, but `RetryRequestError` only disregards the `maxRequestRetries` check, but **is still incrementing the retry count**. 

I'm imagining a doomsday scenario where you first get 5 proxy errors, then some (retriable) errors in your `requestHandler` once you finally accessed the page -  but because the request has failed so many times on proxies, it won't retry anymore and will fail. 

I suppose changing the `RetryRequestError` behaviour would be breaking though, so this is IMO fine? 🤷🏼

Reply:
Maybe we could have a parameter in `RetryRequestError` to specify "max explicit retries" for this request, by default it would retry forever (as now), but you could do `new RetryRequestError(10)` or something like that to limit the explicit retries.

Reply:
I kind of like it, but I cannot really imagine the implementation 
- Is the parameter (10) in regard to the "global" (request-wide) retry count? (then it's not really usable, as different errors on this request can exhaust this shared limit) 
- Is the limit just for this error instance? (that doesn't really match the "functional" way we process requests in Crawlee)
- Or is the limit just for this spot where we call this specific error constructor? (I'd love to have something like this, but sounds like antipattern with analyzing call stacks)

Or maybe I am just not getting something, sorry... I'm interested in this approach, anything to deduplicate the logic, feel free to elaborate more :)

Reply:
Maybe we don't have to bind this to the error instance at all, if we have this as a config option, we could respect that when we see `RetryRequestError` and ensure given request is never retried more than that number (again, only checked for the `RetryRequestError` errors).

Reply:
Idk, I'm assuming `RetryRequestError` is something the user throws themselves, I don't really want to affect the session rotation logic with user code and vice versa. Also, limiting `RetryRequestError`s would be borderline breaking, as they get retried indefinitely now. 

I've now added a new `maxSessionRotations` option, to me it feels kinda reasonable now, let's discuss that?